docs: explain lazy execution in architecture flow (#6994)
Eventual-Inc/Daft: docs/architecture/index.md

|   Output schema = a#Int64
```

## Lazy Evaluation Boundary

Daft DataFrames are lazy until a terminal API asks for materialized results. Python owns the user-facing `DataFrame` object, but the DataFrame stores a Rust-backed `LogicalPlanBuilder` that represents the work to perform. Transformations such as `select`, `filter`, `join`, and `groupby` do not execute data; they call into the Rust logical-plan builder and return a new DataFrame wrapping the updated plan.

Execution begins when a materializing operation such as `collect()`, `show()`, `to_arrow()`, or a write method calls the active runner. At that point Daft freezes the relevant execution context, optimizes the logical plan, lowers it into a physical plan, and streams materialized partitions back through the runner.

The high-level boundary looks like this:

```text
Python DataFrame API
  -> Rust LogicalPlanBuilder
  -> optimized Rust logical plan
  -> local or distributed physical plan
  -> native or Ray execution
  -> materialized partitions
```

> 

Reviewer: Agree that we should be explicit about _when_ a query is executed, but I don't think it needs a separate section in the architecture docs for it.

Right now the top of the architecture doc says "The architecture of Daft consists of 3 main layers: API, Planning, and Execution."

I think the explicit call out for lazy execution actually can just be folded in nicely into these 3 stages. first in the API, when we introduce operators we can already just call it out there that execution doesn't happen yet, we are simply creating a plan of execution.

And then in the execution section, we can call it out that execution is triggered on a 'materializing operation' and then define what that means and give examples.

Reviewer: Side note: While the header states that "The architecture of Daft consists of 3 main layers: API, Planning, and Execution.", the 3 sections of this doc is actually "api, optimization, execution", which is inconsistent. I suggest either consolidating around "api, optimization, execution", or "planning, optimization, execution". My preference is the latter, which makes it explicit that there is a "planning" stage that does not execute.

## Optimization

> Executing the logical plan as written would be correct but often inefficient. Daft applies both rule-based and cost-based optimization after execution is requested and before data processing begins.

Reviewer: ```suggestion
Executing the logical plan as written would be correct but often inefficient. Daft applies both rule-based and cost-based optimization once execution is requested.
```

## Execution

Execution is triggered by a materializing operation such as `collect()`, `show()`, `to_arrow()`, and write methods such as `write_parquet()`.

> Once optimization completes, the optimized logical plan is lowered into a physical plan and executed by either the native runner (single-machine) or the Ray runner (distributed).

Reviewer: ```suggestion
Once optimization completes, the plan is executed by either the native runner (single-machine) or the Ray runner (distributed).
```

No need to introduce the idea of 'lowering into physical plan' since we never talk about what a physical plan is
